### PR TITLE
Add ViewState context and warning message for story URLs

### PR DESCRIPTION
**Pull request by nf-s**
### Add ViewState context and warning message for story URLs

Fixes https://github.com/TerriaJS/vic-digital-twin/issues/333

* Add `ViewState` React context provider to `StandardUserInterface` - instead of passing `viewState` or `terria` props through components, please use
  * `useViewState` hook
  * `withViewState` HOC
* Move `GlobalTerriaStyles` from `StandardUserInterface` to separate file
* Add `ExternalLinkWithWarning` component - this will replace all URLs in story body and add a warning message when URLs are clicked on.
  
image

### Test me
  
- http://ci.terria.io/url-warning/#start=%7B%22version%22%3A%228.0.0%22%2C%22initSources%22%3A%5B%7B%22stratum%22%3A%22user%22%2C%22models%22%3A%7B%7D%2C%22workbench%22%3A%5B%5D%2C%22timeline%22%3A%5B%5D%2C%22initialCamera%22%3A%7B%22west%22%3A105.1183368477886%2C%22south%22%3A-39.52478864283954%2C%22east%22%3A161.8816636411217%2C%22north%22%3A-9.054292826636315%2C%22position%22%3A%7B%22x%22%3A-6733654.97437972%2C%22y%22%3A7095791.782500638%2C%22z%22%3A-4863110.017932153%7D%2C%22direction%22%3A%7B%22x%22%3A0.6155631636719867%2C%22y%22%3A-0.6486682277326119%2C%22z%22%3A0.44756175200788473%7D%2C%22up%22%3A%7B%22x%22%3A-0.30808117990014067%2C%22y%22%3A0.32464982435187667%2C%22z%22%3A0.8942530280293342%7D%7D%2C%22homeCamera%22%3A%7B%22west%22%3A109%2C%22south%22%3A-45%2C%22east%22%3A158%2C%22north%22%3A-8%7D%2C%22viewerMode%22%3A%223dSmooth%22%2C%22showSplitter%22%3Afalse%2C%22splitPosition%22%3A0.5%2C%22settings%22%3A%7B%22baseMaximumScreenSpaceError%22%3A2%2C%22useNativeResolution%22%3Atrue%2C%22alwaysShowTimeline%22%3Afalse%2C%22baseMapId%22%3A%22basemap-positron%22%2C%22terrainSplitDirection%22%3A0%2C%22depthTestAgainstTerrainEnabled%22%3Afalse%7D%2C%22stories%22%3A%5B%7B%22title%22%3A%22Test+scene+with+a+URL%22%2C%22text%22%3A%22%3Cp%3EHere+is+a+URL%3C%2Fp%3E%3Cp%3E%3Cbr%3E%3C%2Fp%3E%3Cp%3E%3Ca+href%3D%5C%22http%3A%2F%2Fgoogle.com%5C%22+data-mce-href%3D%5C%22http%3A%2F%2Fgoogle.com%5C%22%3Ehttp%3A%2F%2Fgoogle.com%3C%2Fa%3E%3C%2Fp%3E%3Cp%3E%3Cbr%3E%3C%2Fp%3E%3Cp%3E%3Cbr+data-mce-bogus%3D%5C%221%5C%22%3E%3C%2Fp%3E%22%2C%22id%22%3A%2217e93465-21c2-4a41-8900-81d4abba69c8%22%2C%22shareData%22%3A%7B%22version%22%3A%228.0.0%22%2C%22initSources%22%3A%5B%7B%22stratum%22%3A%22user%22%2C%22models%22%3A%7B%7D%2C%22workbench%22%3A%5B%5D%2C%22timeline%22%3A%5B%5D%2C%22initialCamera%22%3A%7B%22west%22%3A103.79328350307286%2C%22south%22%3A-43.76783640119821%2C%22east%22%3A163.2067169859072%2C%22north%22%3A-4.746275947047687%2C%22position%22%3A%7B%22x%22%3A-6733654.97437972%2C%22y%22%3A7095791.782500638%2C%22z%22%3A-4863110.017932153%7D%2C%22direction%22%3A%7B%22x%22%3A0.6155631636719867%2C%22y%22%3A-0.6486682277326119%2C%22z%22%3A0.44756175200788473%7D%2C%22up%22%3A%7B%22x%22%3A-0.30808117990014067%2C%22y%22%3A0.32464982435187667%2C%22z%22%3A0.8942530280293342%7D%7D%2C%22homeCamera%22%3A%7B%22west%22%3A109%2C%22south%22%3A-45%2C%22east%22%3A158%2C%22north%22%3A-8%7D%2C%22viewerMode%22%3A%223dSmooth%22%2C%22showSplitter%22%3Afalse%2C%22splitPosition%22%3A0.4999%2C%22settings%22%3A%7B%22baseMaximumScreenSpaceError%22%3A2%2C%22useNativeResolution%22%3Atrue%2C%22alwaysShowTimeline%22%3Afalse%2C%22baseMapId%22%3A%22basemap-positron%22%2C%22terrainSplitDirection%22%3A0%2C%22depthTestAgainstTerrainEnabled%22%3Afalse%7D%7D%5D%7D%7D%5D%7D%5D%7D
- Start story
- Clink on google.com link

### Checklist

-   [ ] There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)
-   [ ] I've updated relevant documentation in `doc/`.
-   [x] I've updated CHANGES.md with what I changed.
-   [x] I've provided instructions in the PR description on how to test this PR.

### CODE REVIEW

File: lib/ReactViews/Custom/ExternalLink.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  attributes: AnchorHTMLAttributes<HTMLAnchorElement>;
+  children: React.ReactNode;
+}
+
+export const ExternalLinkWithWarning: React.FC<Props> = (props: {
+  attributes: AnchorHTMLAttributes<HTMLAnchorElement>;
+  children: React.ReactNode;
+}) => {
+  const viewState = useViewState();
+  const [t] = useTranslation();
```
Comment (tephenavies) [inline lib/ReactViews/Custom/ExternalLink.tsx]: Prefer 
```suggestion
  const {t} = useTranslation();
```

File: lib/ReactViews/ExplorerWindow/ExplorerWindow.tsx [hunk 2]
```diff
@@ -2,45 +2,47 @@ import { action } from "mobx";
...
-  terria: ViewState["terria"];
 }
 
-export default observer<React.FC<IProps>>(function ExplorerWindow(props) {
-  const onClose = action(() => {
-    props.viewState.closeCatalog();
-    props.viewState.switchMobileView("nowViewing");
-  });
+export default withViewState(
+  observer<React.FC<IProps>>(function ExplorerWindow(props) {
```
Comment (tephenavies) [inline lib/ReactViews/ExplorerWindow/ExplorerWindow.tsx]: Can this be done with `useViewState` to avoid so much indent nesting?

Comment (tephenavies) [reply lib/ReactViews/ExplorerWindow/ExplorerWindow.tsx]: Or wrap the FC in HOCs at the bottom of the file

File: lib/ReactViews/Map/Panels/SharePanel/Print/PrintView.tsx [hunk 3]
```diff
@@ -2,9 +2,11 @@ import DOMPurify from "dompurify";
 import React, { useEffect, useRef, useState } from "react";
 import ReactDOM from "react-dom";
 import { StyleSheetManager, ThemeProvider } from "styled-components";
-import Terria from "../../../../../Models/Terria";
-import ViewState from "../../../../../ReactViewModels/ViewState";
 import { terriaTheme } from "../../../../StandardUserInterface/StandardTheme";
+import {
+  WithViewState,
+  withViewState
```
Comment (tephenavies) [inline lib/ReactViews/Map/Panels/SharePanel/Print/PrintView.tsx]: Should this use `useViewState`?

File: lib/ReactViews/Mobile/MobileHeader.jsx [hunk 4]
```diff
@@ -12,280 +11,267 @@ import { RawButton } from "../../Styled/Button";
...
+              <div
+                className={Styles.groupRight}
+                css={`
+                  background-color: ${p => p.theme.dark};
+                `}
+              >
+                <button
+                  type="button"
+                  className={Styles.btnAdd}
+                  onClick={this.onMobileDataCatalogClicked}
```
Comment (tephenavies) [inline lib/ReactViews/Mobile/MobileHeader.jsx]: Function needs binding.

I hadn't thought of it before, but this is a good reason to strongly discourage class components.

File: lib/ReactViews/Mobile/MobileHeader.jsx [hunk 5]
```diff
@@ -12,280 +11,267 @@ import { RawButton } from "../../Styled/Button";
...
+                    styledHeight="20px"
+                  />
+                </button>
+                <If condition={nowViewingLength > 0}>
                   <button
                     type="button"
-                    className={Styles.btnAdd}
-                    onClick={this.onMobileDataCatalogClicked}
+                    className={Styles.btnNowViewing}
+                    onClick={this.onMobileNowViewingClicked}
```
Comment (tephenavies) [inline lib/ReactViews/Mobile/MobileHeader.jsx]: Also needs binding

File: lib/ReactViews/Mobile/MobileHeader.jsx [hunk 6]
```diff
@@ -12,280 +11,267 @@ import { RawButton } from "../../Styled/Button";
...
-            allBaseMaps={this.props.allBaseMaps}
-            terria={this.props.terria}
-            showFeedback={!!this.props.terria.configParameters.feedbackUrl}
-          />
-          {/* Don't show mobile modal window if user is currently interacting
+                </If>
+                <button
+                  className={Styles.btnSearch}
+                  type="button"
+                  onClick={this.showSearch}
```
Comment (tephenavies) [inline lib/ReactViews/Mobile/MobileHeader.jsx]: Binding

File: lib/ReactViews/Mobile/MobileHeader.jsx [hunk 7]
```diff
@@ -12,280 +11,267 @@ import { RawButton } from "../../Styled/Button";
...
+              <div className={Styles.formSearchData}>
+                <Choose>
+                  <When condition={searchState.showMobileLocationSearch}>
+                    <SearchBox
+                      searchText={searchState.locationSearchText}
+                      onSearchTextChanged={this.changeLocationSearchText}
+                      onDoSearch={this.searchLocations}
+                      placeholder={t("search.placeholder")}
+                      alwaysShowClear={true}
+                      onClear={this.closeLocationSearch}
```
Comment (tephenavies) [inline lib/ReactViews/Mobile/MobileHeader.jsx]: Binding needed for all props

File: lib/ReactViews/Mobile/MobileHeader.jsx [hunk 8]
```diff
@@ -12,280 +11,267 @@ import { RawButton } from "../../Styled/Button";
...
+                      autoFocus={true}
+                    />
+                  </When>
+                  <When condition={searchState.showMobileCatalogSearch}>
+                    <SearchBox
+                      searchText={searchState.catalogSearchText}
+                      onSearchTextChanged={this.changeCatalogSearchText}
+                      onDoSearch={this.searchCatalog}
+                      placeholder={t("search.searchCatalogue")}
+                      onClear={this.closeCatalogSearch}
```
Comment (tephenavies) [inline lib/ReactViews/Mobile/MobileHeader.jsx]: Function binding

File: lib/ReactViews/StandardUserInterface/StandardUserInterface.tsx [hunk 9]
```diff
@@ -217,208 +147,173 @@ const StandardUserInterface = observer<React.FC<StandardUserInterfaceProps>>(
...
-          )}
-          <div
-            className={Styles.uiRoot}
-            css={`
-              ${props.viewState.disclaimerVisible && `filter: blur(10px);`}
-            `}
-            onDragOver={handleDragOver}
-          >
+      <ViewStateProvider viewState={props.viewState}>
+        <ThemeProvider theme={mergedTheme}>
```
Comment (tephenavies) [inline lib/ReactViews/StandardUserInterface/StandardUserInterface.tsx]: I wonder if we should break out providers into their own file. Then we'll feel more free to add more providers and it won't pollute StandardUserInterface.

Comment (tephenavies) [reply lib/ReactViews/StandardUserInterface/StandardUserInterface.tsx]: Up to here with my review.

**Review by tephenavies — CHANGES_REQUESTED** [review]
some changes needed. not yet finished review

File: lib/ReactViews/StandardUserInterface/StandardUserInterface.tsx (hunk 9 above)
Comment (nf-s) [reply lib/ReactViews/StandardUserInterface/StandardUserInterface.tsx]: done!

**Review by tephenavies — APPROVED** [review]

**Pull request merged**